contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1375 @@
+        let msg = ExecuteMsg::EnableMint {};
+        execute(deps.as_mut(), env.clone(), info, msg).unwrap();
+
+        env.block.time = Timestamp::from_seconds(EVENT_END_SECONDS);
+
+        let msg = ExecuteMsg::Mint {};
+        let info = mock_info(USER, &vec![]);
+        let result = execute(deps.as_mut(), env.clone(), info, msg);
+
+        // Event is not started
+        assert_eq!(
+            ContractError::EventTerminated {
+                current_time: env.block.time,
+                end_time: Timestamp::from_seconds(EVENT_END_SECONDS),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn mint_without_permissions_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(USER, &vec![]);
+
+        do_instantiate(deps.as_mut());
+
+        // Change current time to event start
+        env.block.time = Timestamp::from_seconds(EVENT_START_SECONDS);
+
+        let msg = ExecuteMsg::Mint {};
+        let result = execute(deps.as_mut(), env.clone(), info, msg);
+
+        // Event is not started
+        assert_eq!(ContractError::MintDisabled {}, result.unwrap_err());
+    }
+
+    #[test]
+    fn mint_limit() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(ADMIN, &vec![]);
+
+        do_instantiate(deps.as_mut());
+
+        // Change current time to event start
+        env.block.time = Timestamp::from_seconds(EVENT_START_SECONDS);
+
+        execute(deps.as_mut(), env.clone(), info, ExecuteMsg::EnableMint {}).unwrap();
+
+        let info = mock_info(USER, &vec![]);
+        // Mint the first poap
+        execute(
+            deps.as_mut(),
+            env.clone(),
+            info.clone(),
+            ExecuteMsg::Mint {},
+        )
+        .unwrap();
+        // Mint the second poap
+        execute(
+            deps.as_mut(),
+            env.clone(),
+            info.clone(),
+            ExecuteMsg::Mint {},
+        )
+        .unwrap();
+
+        let response = execute(
+            deps.as_mut(),
+            env.clone(),
+            info.clone(),
+            ExecuteMsg::Mint {},
+        );
+        assert_eq!(
+            ContractError::MaxPerAddressLimitExceeded {
+                recipient_addr: USER.to_string()
+            },
+            response.unwrap_err()
+        );
+
+        // Ensure that mint to also fails when minting for the user
+        let info = mock_info(ADMIN, &vec![]);
+        let response = execute(
+            deps.as_mut(),
+            env.clone(),
+            info,
+            ExecuteMsg::MintTo {
+                recipient: USER.to_string(),
+            },
+        );
+
+        assert_eq!(
+            ContractError::MaxPerAddressLimitExceeded {
+                recipient_addr: USER.to_string()
+            },
+            response.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn mint_to_limit() {
```

Comment:
It could be removed since it is duplicated with `mint_limit`

Reply:
If we remove this test we don't ensure that the `MintTo` message don't allow to mint more then `per_address_limit` poap.
This test can be renamed into `mint_to_out_of_max_amount_error` and add the missing integration test `mint_to_with_permission_properly`.

Reply:
Yes, my bad, I forgot that `mint_to` and `mint` are different messages.